src/ParseFasm.cpp

```diff
@@ -0,0 +1,508 @@
+                                bits += 3;
+                                if (bits >= Num::kWidth) {
+                                        data << Num(word >>
+                                                    (bits - Num::kWidth));
+                                        word >>= Num::kWidth;
+                                        bits -= Num::kWidth;
+                                }
+                        }
+                        it++;
+                }
+                assert(!word);
+                return withHeader(TAG('o', octal), data.str());
+        }
+
+        /// A collection of annotations. { ... }
+        /// Tags: annotations ({)
+        virtual Any visitAnnotations(
+            FasmParser::AnnotationsContext* context) override {
+                std::ostringstream data;
+                for (auto& a : context->annotation()) {
+                        data << visit(a).as<std::string>();
+                }
+                return withHeader(TAG('{', annotations), data.str());
+        }
+
+        /// An annotation: x = "y"
+        /// Tags: annotation (a), annotation name (.), annotation value (=)
+        virtual Any visitAnnotation(
+            FasmParser::AnnotationContext* context) override {
+                std::ostringstream data;
+                data << Str(TAG('.', annotation_name),
+                            context->ANNOTATION_NAME()->getText());
+                if (context->ANNOTATION_VALUE()) {
+                        std::string value =
+                            context->ANNOTATION_VALUE()->getText();
+                        value.erase(0, 1);  /// Convert "value" -> value
+                        value.pop_back();
+                        data << Str(TAG('=', annotation_value), value);
+                }
+                return withHeader(TAG('a', annotation), data.str());
+        }
+
+       private:
+        std::ostream& out;
+};
+
+/// Common portion of 'from_string' and 'from_file'.
+/// Consumes an input stream and produces an output stream.
+static void parse_fasm(std::istream& in, std::ostream& out) {
+        ANTLRInputStream stream(in);
+        FasmLexer lexer(&stream);
+        CommonTokenStream tokens(&lexer);
+        FasmParser parser(&tokens);
+        parser.setErrorHandler(std::make_shared<BailErrorStrategy>());
+        auto* tree = parser.fasmFile();
+        FasmParserBaseVisitor(out).visit(tree);
+}
+
+/// Parse the given input string, returning output.
+/// Use hex mode (see above) if hex is true.
+/// Use a callback to avoid copying the result.
+void from_string(const char* in, bool hex, void (*ret)(const char* str)) {
+        hex_mode = hex;
+        std::istringstream input(in);
+        std::ostringstream output;
+
+        try {
+                parse_fasm(input, output);
+                output.put(0);
+                std::string result = output.str();
+                ret(result.c_str());
+        } catch (ParseCancellationException e) {
```

Comment:
This feels weird.  Can you put a small comment on why this `try`/`catch` is here?

Reply:
Done. It's to catch parse failures, in which the callback is not called back.

Reply:
Should there be an error callback with more information?  Maybe a line number, etc?  Parsing errors with textx usually gave enough information to determine how to resolve the problem.